## Part Time Data Science 07 class: Phase 3 Project
### Project by: Nicholas Kirui
### Date of Submission: 01 September 2024

## Import libraries


In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline

#  Import dataset


In [2]:
data = './Data/bigml.csv'

df = pd.read_csv(data)

# Exploratory data analysis

Explore the data to gain insights about the data. 

In [3]:
# preview the dataset
df.head(5)

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [4]:
# view dimensions of dataset
df.shape

(3333, 21)

Dataset has 21 variables and 3333 instances

### Identify the columns in the dataset

In [5]:
# review the column names
col_names = df.columns

col_names

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')

In [6]:
# view summary of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

### Types of variables
In this section, I segregate the dataset into categorical and numerical variables. 
- There are a mixture of categorical and numerical variables in the dataset. 
- Categorical variables have data type object
- Numerical variables have data type float64

#### First of all, I will find categorical variables

In [7]:
# find categorical variables

categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :', categorical)

There are 4 categorical variables

The categorical variables are : ['state', 'phone number', 'international plan', 'voice mail plan']


In [8]:
# view the categorical variables

df[categorical].head()

state phone number international plan voice mail plan
0    KS     382-4657                 no             yes
1    OH     371-7191                 no             yes
2    NJ     358-1921                 no              no
3    OH     375-9999                yes              no
4    OK     330-6626                yes              no

### Summary of categorical variables

- There are 4 categorical variables. These are given by:
        -`state`, `phone number`, `international plan`,  and  `voice mail plan`


- There are two binary categorical variables:
        - `international plan`,  and  `voice mail plan`


## Explore problems within categorical variables

### Missing values in categorical variables

In [9]:
# check missing values in categorical variables

df[categorical].isnull().sum()

state                 0
phone number          0
international plan    0
voice mail plan       0
dtype: int64

There are no missing values in categorical variables

### Frequency counts of categorical variables

check the frequency counts of categorical variables.

In [14]:
# view frequency of categorical variables - state

print(df['state'].value_counts())

WV    106
MN     84
NY     83
AL     80
OR     78
OH     78
WI     78
VA     77
WY     77
CT     74
ID     73
VT     73
MI     73
UT     72
TX     72
IN     71
MD     70
KS     70
NJ     68
MT     68
NC     68
NV     66
WA     66
CO     66
RI     65
MS     65
MA     65
AZ     64
MO     63
FL     63
ND     62
ME     62
NM     62
OK     61
NE     61
DE     61
SC     60
SD     60
KY     59
IL     58
NH     56
AR     55
DC     54
GA     54
HI     53
TN     53
AK     52
LA     51
PA     45
IA     44
CA     34
Name: state, dtype: int64


In [15]:
# view frequency of categorical variables: phone number

print(df['phone number'].value_counts())

388-8282    1
393-8762    1
383-9355    1
336-6085    1
335-5882    1
           ..
376-1641    1
344-6404    1
334-9182    1
409-1244    1
365-8779    1
Name: phone number, Length: 3333, dtype: int64


In [13]:
# view frequency of categorical variables: international plan

print(df['international plan'].value_counts())

no     3010
yes     323
Name: international plan, dtype: int64


In [12]:
# view frequency of categorical variables: voice mail plan

print(df['voice mail plan'].value_counts())

no     2411
yes     922
Name: voice mail plan, dtype: int64


### View the frequency distribution of categorical variables

In [16]:
# view frequency distribution of categorical variables: state

print(df['state'].value_counts()/np.float64(len(df)))

WV    0.031803
MN    0.025203
NY    0.024902
AL    0.024002
OR    0.023402
OH    0.023402
WI    0.023402
VA    0.023102
WY    0.023102
CT    0.022202
ID    0.021902
VT    0.021902
MI    0.021902
UT    0.021602
TX    0.021602
IN    0.021302
MD    0.021002
KS    0.021002
NJ    0.020402
MT    0.020402
NC    0.020402
NV    0.019802
WA    0.019802
CO    0.019802
RI    0.019502
MS    0.019502
MA    0.019502
AZ    0.019202
MO    0.018902
FL    0.018902
ND    0.018602
ME    0.018602
NM    0.018602
OK    0.018302
NE    0.018302
DE    0.018302
SC    0.018002
SD    0.018002
KY    0.017702
IL    0.017402
NH    0.016802
AR    0.016502
DC    0.016202
GA    0.016202
HI    0.015902
TN    0.015902
AK    0.015602
LA    0.015302
PA    0.013501
IA    0.013201
CA    0.010201
Name: state, dtype: float64


In [17]:
# view frequency distribution of categorical variables: phone number

print(df['phone number'].value_counts()/np.float64(len(df)))

388-8282    0.0003
393-8762    0.0003
383-9355    0.0003
336-6085    0.0003
335-5882    0.0003
             ...  
376-1641    0.0003
344-6404    0.0003
334-9182    0.0003
409-1244    0.0003
365-8779    0.0003
Name: phone number, Length: 3333, dtype: float64


In [18]:
# view frequency distribution of categorical variables: international plan

print(df['international plan'].value_counts()/np.float64(len(df)))

no     0.90309
yes    0.09691
Name: international plan, dtype: float64


In [19]:
# view frequency distribution of categorical variables: voice mail plan

print(df['voice mail plan'].value_counts()/np.float64(len(df)))

no     0.723372
yes    0.276628
Name: voice mail plan, dtype: float64


### Number of labels: cardinality

The number of labels within a categorical variable is known as **cardinality**. A high number of labels within a variable is known as **high cardinality**. High cardinality may pose some serious problems in the machine learning model. So, check for high cardinality.

In [20]:
# check for cardinality in categorical variables

for var in categorical:
        print(var, ' contains ', len(df[var].unique()), ' labels')

state  contains  51  labels
phone number  contains  3333  labels
international plan  contains  2  labels
voice mail plan  contains  2  labels


- phone number has the same number of labels as there are instances

### Explore Categorical Variables

Now, explore the categorical variables one by one. 

In [21]:
# find categorical variables
categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :', categorical)

There are 4 categorical variables

The categorical variables are : ['state', 'phone number', 'international plan', 'voice mail plan']


### Explore the `state` variable

In [23]:
# print number of labels in state variable
print('state contains', len(df["state"].unique()), 'labels')

state contains 51 labels


In [24]:
# check labels in state variable
df.state.unique()

array(['KS', 'OH', 'NJ', 'OK', 'AL', 'MA', 'MO', 'LA', 'WV', 'IN', 'RI',
       'IA', 'MT', 'NY', 'ID', 'VT', 'VA', 'TX', 'FL', 'CO', 'AZ', 'SC',
       'NE', 'WY', 'HI', 'IL', 'NH', 'GA', 'AK', 'MD', 'AR', 'WI', 'OR',
       'MI', 'DE', 'UT', 'CA', 'MN', 'SD', 'NC', 'WA', 'NM', 'NV', 'DC',
       'KY', 'ME', 'MS', 'TN', 'PA', 'CT', 'ND'], dtype=object)

In [25]:
# check frequency distribution of values in state variable

df["state"].value_counts()

WV    106
MN     84
NY     83
AL     80
OR     78
OH     78
WI     78
VA     77
WY     77
CT     74
ID     73
VT     73
MI     73
UT     72
TX     72
IN     71
MD     70
KS     70
NJ     68
MT     68
NC     68
NV     66
WA     66
CO     66
RI     65
MS     65
MA     65
AZ     64
MO     63
FL     63
ND     62
ME     62
NM     62
OK     61
NE     61
DE     61
SC     60
SD     60
KY     59
IL     58
NH     56
AR     55
DC     54
GA     54
HI     53
TN     53
AK     52
LA     51
PA     45
IA     44
CA     34
Name: state, dtype: int64

In [46]:
# let's do One Hot Encoding of state variable
# get k-1 dummy variables after One Hot Encoding 
# also add an additional dummy variable to indicate there was missing data
# preview the dataset with head() method

pd.get_dummies(df['state'], drop_first=True, dummy_na=True,dtype='int').head()

AL  AR  AZ  CA  CO  CT  DC  DE  FL  GA  ...  TN  TX  UT  VA  VT  WA  WI  \
0   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
1   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
2   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
3   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   

   WV  WY  NaN  
0   0   0    0  
1   0   0    0  
2   0   0    0  
3   0   0    0  
4   0   0    0  

[5 rows x 51 columns]

In [47]:
# sum the number of 1s per boolean variable over the rows of the dataset
# it will tell us how many observations we have for each category

pd.get_dummies(df['state'], drop_first=True, dummy_na=True).sum(axis=0)

AL      80
AR      55
AZ      64
CA      34
CO      66
CT      74
DC      54
DE      61
FL      63
GA      54
HI      53
IA      44
ID      73
IL      58
IN      71
KS      70
KY      59
LA      51
MA      65
MD      70
ME      62
MI      73
MN      84
MO      63
MS      65
MT      68
NC      68
ND      62
NE      61
NH      56
NJ      68
NM      62
NV      66
NY      83
OH      78
OK      61
OR      78
PA      45
RI      65
SC      60
SD      60
TN      53
TX      72
UT      72
VA      77
VT      73
WA      66
WI      78
WV     106
WY      77
NaN      0
dtype: int64

### Explore `phone number` variable

In [27]:
# print number of labels in phone number variable
print('phone number contains', len(df["phone number"].unique()), 'labels')

phone number contains 3333 labels


In [31]:
# check labels in phone number variable
df["phone number"].unique()

array(['382-4657', '371-7191', '358-1921', ..., '328-8230', '364-6381',
       '400-4344'], dtype=object)

In [35]:
# check frequency distribution of values in phone number variable

df["phone number"].value_counts()

388-8282    1
393-8762    1
383-9355    1
336-6085    1
335-5882    1
           ..
376-1641    1
344-6404    1
334-9182    1
409-1244    1
365-8779    1
Name: phone number, Length: 3333, dtype: int64

### Explore `intenational plan` variable

In [28]:
# print number of labels in international plan variable
print('international plan contains', len(df["international plan"].unique()), 'labels')

international plan contains 2 labels


In [32]:
# check labels in international plan variable
df["international plan"].unique()

array(['no', 'yes'], dtype=object)

In [36]:
# check frequency distribution of values in international plan variable

df["international plan"].value_counts()

no     3010
yes     323
Name: international plan, dtype: int64

In [45]:
# let's do One Hot Encoding of international plan variable
# get k-1 dummy variables after One Hot Encoding 
# also add an additional dummy variable to indicate there was missing data
# preview the dataset with head() method

pd.get_dummies(df['international plan'], drop_first=True, dummy_na=True,dtype='int').head()

yes  NaN
0    0    0
1    0    0
2    0    0
3    1    0
4    1    0

In [39]:
# sum the number of 1s per boolean variable over the rows of the dataset
# it will tell us how many observations we have for each category

pd.get_dummies(df['international plan'], drop_first=True, dummy_na=True).sum(axis=0)

yes    323
NaN      0
dtype: int64

### Explore `voice mail plan` variable

In [29]:
# print number of labels in voice mail plan variable
print('voice mail plan contains', len(df["voice mail plan"].unique()), 'labels')

voice mail plan contains 2 labels


In [34]:
# check labels in voice mail plan variable
df["voice mail plan"].unique()

array(['yes', 'no'], dtype=object)

In [37]:
# check frequency distribution of values in voice mail plan variable

df["voice mail plan"].value_counts()

no     2411
yes     922
Name: voice mail plan, dtype: int64

In [48]:
# let's do One Hot Encoding of voice mail plan variable
# get k-1 dummy variables after One Hot Encoding 
# also add an additional dummy variable to indicate there was missing data
# preview the dataset with head() method

pd.get_dummies(df['voice mail plan'], drop_first=True, dummy_na=True,dtype='int').head()

yes  NaN
0    1    0
1    1    0
2    0    0
3    0    0
4    0    0

In [50]:
# sum the number of 1s per boolean variable over the rows of the dataset
# it will tell us how many observations we have for each category

pd.get_dummies(df['voice mail plan'], drop_first=True, dummy_na=True).sum(axis=0)

yes    922
NaN      0
dtype: int64